File: Arrays 101/In Place Operations/ReplaceElementsWithGreatestElementOnRightSide.java
// Given an array arr, replace every element in that array with the greatest element among the elements to its right, and replace the last element with -1.

// After doing so, return the array.

class Solution {
    public int[] replaceElements(int[] arr) {
        // input : integer array
        // output: integer array

        // edge case: if length of array is less than 2 ([])
        if (arr.length == 0) {
            return arr;
        }

        if (arr.length == 1) {
            arr[0] = -1;
        }

        // create temp variable to store current element
        int tempVal;
        // set max value to last element
        int maxVal = arr[arr.length - 1];

        // set last element to -1
        arr[arr.length - 1] = -1;

        // iterate from second to last element
        for (int i = arr.length - 2; i >= 0; i--) {

            // set temp value to current value
            tempVal = arr[i];

            // set the current value to max value
            arr[i] = maxVal;

            // check if temp value > max value
            if (tempVal > maxVal) {

                // set temp value to max value
                maxVal = tempVal;
            }
        }
        // return the array
        return arr;
    }
}

File: Arrays 101/In Place Operations/Replace Elements With Greatest Element On Right Side/ReplaceElementsWithGreatestElementOnRightSide.py
# Given an array arr, replace every element in that array with the greatest element among the elements to its right, and replace the last element with -1.

# After doing so, return the array.

class Solution:
    def replaceElements(self, arr: List[int]) -> List[int]:
        # input : int array
        # output: int array(in-place)

        # edge case: when length of arr == 0
        if len(arr) == 0:

            return arr

        # edge case: when length of arr == 1
        if len(arr) == 1:

            return [-1]

        # create var for max value: set it to last element
        maxVal = arr[-1]

        # set last element to -1
        arr[-1] = -1

        # iterate from the second to last element
        for i in range(len(arr) - 2, -1, -1):

            # create a temp variable to store current element
            temp = arr[i]

            # replace current element with the max value
            arr[i] = maxVal

            # if the temp variable > max value
            if temp > maxVal:

                # replace the max value with temp variable
                maxVal = temp

        # return arr
        return arr
